game-core/src/main/java/games/strategy/triplea/ai/pro/util/MoveBatcher.java

```diff
@@ -0,0 +1,161 @@
+package games.strategy.triplea.ai.pro.util;
+
+import games.strategy.engine.data.Route;
+import games.strategy.engine.data.Unit;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * MoveBatcher implements an algorithm that batches AI moves together, reducing the number of
+ * distinct moves the AI does (and thus the visual delay) by combining moves together.
+ */
+public class MoveBatcher {
+  private static class Move {
+    private final ArrayList<Unit> units;
+    private final Route route;
+    private final @Nullable HashMap<Unit, Unit> unitsToTransports;
+
+    private Move(
+        final ArrayList<Unit> units,
+        final Route route,
+        final @Nullable HashMap<Unit, Unit> unitsToTransports) {
+      this.units = units;
+      this.route = route;
+      this.unitsToTransports = unitsToTransports;
+    }
+
+    Move(final ArrayList<Unit> units, final Route route) {
+      this(units, route, null);
```

Comment:
To be clear, null entries is how the three parallel lists in ProMoveUtils are handling transports already. So changing it here would require a bigger refactoring. But the right bigger refactoring is to get rid of the three parallel lists in the first place. We agreed that that should be done in a separate follow-up PR.

Reply:
Okay, I like how you annotated the nullables later so at least we have that safe-guard. I'm sure there is a lot of refactoring to be done. I suppose it's of help if the locations for update are marked so that it is communicated for review and other developers if they happen to read the code, and so we don't forget.